pkg/agent/memberlist/cluster_test.go

```diff
@@ -0,0 +1,745 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package memberlist
+
+import (
+	"context"
+	"crypto/rand"
+	"fmt"
+	"net"
+	"reflect"
+	"testing"
+	"time"
+
+	"github.com/golang/groupcache/consistenthash"
+	"github.com/hashicorp/memberlist"
+	"github.com/stretchr/testify/assert"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes/fake"
+	"k8s.io/client-go/tools/cache"
+
+	"antrea.io/antrea/pkg/agent/config"
+	"antrea.io/antrea/pkg/apis"
+	crdv1a2 "antrea.io/antrea/pkg/apis/crd/v1alpha2"
+	fakeversioned "antrea.io/antrea/pkg/client/clientset/versioned/fake"
+	crdinformers "antrea.io/antrea/pkg/client/informers/externalversions"
+)
+
+func TestCluster_Run(t *testing.T) {
+	localNodeName := "test_memberlist_node"
+	nodeConfig := &config.NodeConfig{
+		Name:       localNodeName,
+		NodeIPAddr: &net.IPNet{IP: net.IPv4(127, 0, 0, 1), Mask: net.IPv4Mask(255, 255, 255, 255)},
+	}
+
+	localNode := &v1.Node{
+		ObjectMeta: metav1.ObjectMeta{Name: localNodeName},
+		Status:     v1.NodeStatus{Addresses: []v1.NodeAddress{{Type: v1.NodeInternalIP, Address: "127.0.0.1"}}},
+	}
+
+	eip := &crdv1a2.ExternalIPPool{
+		TypeMeta:   metav1.TypeMeta{Kind: "CustomResourceDefinition"},
+		ObjectMeta: metav1.ObjectMeta{Name: "fake_ip_pool"},
+		Spec:       crdv1a2.ExternalIPPoolSpec{},
+	}
+	eip1 := &crdv1a2.ExternalIPPool{
+		TypeMeta:   metav1.TypeMeta{Kind: "CustomResourceDefinition"},
+		ObjectMeta: metav1.ObjectMeta{Name: "fake_ip_pool1"},
+		Spec:       crdv1a2.ExternalIPPoolSpec{NodeSelector: metav1.LabelSelector{MatchLabels: map[string]string{"env": "pro"}}},
+	}
+
+	fakeEgressIP := "1.1.1.1"
+	eg := &crdv1a2.Egress{
+		Spec: crdv1a2.EgressSpec{ExternalIPPool: eip.Name, EgressIP: fakeEgressIP},
+	}
+
+	testCases := []struct {
+		name                     string
+		egress                   *crdv1a2.Egress
+		externalIPPool           *crdv1a2.ExternalIPPool
+		expectEgressSelectResult bool
+	}{
+		{
+			name:                     "local Node matches ExternalIPPool nodeSelectors",
+			egress:                   eg,
+			externalIPPool:           eip,
+			expectEgressSelectResult: true,
+		},
+		{
+			name:                     "local Node not match ExternalIPPool nodeSelectors",
+			egress:                   eg,
+			externalIPPool:           eip1,
+			expectEgressSelectResult: false,
+		},
+	}
+	for i, tCase := range testCases {
+		t.Run(tCase.name, func(t *testing.T) {
+			stopCh := make(chan struct{})
+			defer close(stopCh)
+
+			port := apis.AntreaAgentClusterPort + i
+
+			cs := fake.NewSimpleClientset()
```

Comment:
I still see the potential for unification. This test file is quite long and I see the same setup code repeated multiple times. Am I missing something/

Reply:
done